RequestContext - Phase 1

problame opened the pull request:
**Note to reviewers**: start with the block comment in `context.rs`

---

Motivation
==========

Layer Eviction Needs Context
----------------------------

Before we start implementing layer eviction, we need to collect some access statistics per layer file or maybe even page. Part of these statistics should be the initiate of a page read request. I.e., was it page_service vs. a background loop, and if the latter, which of them.

Further, it would be nice to learn more about what activity in the pageserver initated an on-demand download of a layer file.
We will use this information to test out layer eviction policies.

Read more about the current plan for layer eviction here: https://github.com/neondatabase/neon/issues/2476#issuecomment-1370822104

task_mgr problems + cancellation + tenant/timeline lifecycle
---------------------------------------------------------------------------------------

Apart from layer eviction, we have long-standing problems with task_mgr, tenant / timeline deletion, and generally, a unified cancellation mechanism.
One approach to solve this is to abandon task_mgr in favor of a mechanism similar to Golang's context.Context, albeit extended to support waiting for completion, and specialized to the needs in the pageserver.

Heikki solves all of the above at once in PR
https://github.com/neondatabase/neon/pull/3228 , which is not yet merged at the time of writing.

What Is This Patch About
========================

This patch addresses the immediate needs for layer eviction by introducing a RequestContext structure (name up for debate) that is plumbed through all the way from the entrypoints of the pageserver (page_service, management API, tenant background loops) all the way down to Timeline::{get,get_reconstruct_data}.

The struct carries a description of the kind of activity that initiated the call (re-using task_mgr::TaskKind here).

Also, it carries the desired on-demand download behavior of the entrypoint. Timeline::get_reconstruct_data can then log the TaskKind that initiated the on-demand download.

I developed this patch by git-checking-out Heikki's big RequestContext PR https://github.com/neondatabase/neon/pull/3228 , then deleting all the functionality that we do not need to address the needs for layer eviction.

This patch is huge, it should be easy to review because almost all of it is plumbing through of the &RequestContext argument.

Also, this patch is low risk because it should not be changing any behavior (apart from DownloadBehavior::Error).

TODO:
- Self-review
- We could reduce the size a bit by removing the various occurences of ``` fn(..., _ctx: &RequestContext) ``` that I left in place from PR 3228.

hlinnaka commented:
> Before we start implementing layer eviction, we need to collect some access statistics per layer file or maybe even page. Part of these statistics should be the initiate of a page read request. I.e., was it page_service vs. a background loop, and if the latter, which of them.

Ah, that's why you wanted to do this PR separately. Makes sense!

hlinnaka reviewed (approved):
RemoteTimelineClient::download_index_file and `download_layer_file` probably should take a RequestContext parameter too. I presume that's where you would actually collect the metrics of who downloads what.

I didn't read this through completely, but `context.rs` looks good to me, and I presume the rest is just a rehash of the changes from PR #3228. Let me know if some particular part here needs review.

SomeoneToIgnore commented on pageserver/src/task_mgr.rs:
```diff
@@ -1,4 +1,6 @@
 //!
+//! TODO need update comment / at least mention context.rs
```
👍

problame replied:
Left `task_mgr.rs` as is, since this patch doesn't change anything about it.

Added a large block comment to `context.rs`, though.

SomeoneToIgnore commented on pageserver/src/tenant/mgr.rs:
```diff
@@ -523,7 +530,10 @@ pub async fn immediate_compact(
-                .compact()
+                .compact(&RequestContext::new(
+                    TaskKind::Compaction,
+                    DownloadBehavior::Warn, /* TODO */
```
Looks like `Download` to me.

problame replied:
Right.

problame replied:
Yep, you're right.

problame commented on pageserver/src/tenant/timeline.rs:
```diff
@@ -939,6 +983,9 @@ impl Timeline {
         }
 
+        let background_ctx =
+            RequestContext::new(TaskKind::WalReceiverManager, DownloadBehavior::Error);
```
`DownloadBehavior::Error` @hlinnaka I copy-pasted this from your PR. Are you sure this works in practice? IIRC walreceiver sometimes needs to access relsize or so to determine whether rel extension is needed.

hlinnaka replied:
Ingesting WAL records indeed needs to access relation sizes, but this context is for the "connection manager". A separate child context is created for each connection which performs the ingestion.

sharnoff commented on pageserver/src/tenant.rs:
```diff
@@ -1101,11 +1115,12 @@ impl Tenant {
     /// This is used to create the initial 'main' timeline during bootstrapping,
     /// or when importing a new base backup. The caller is expected to load an
     /// initial image of the datadir to the new timeline after this.
-    pub fn create_empty_timeline(
+    pub fn create_empty_timeline<'a>(
         &self,
         new_timeline_id: TimelineId,
         initdb_lsn: Lsn,
         pg_version: u32,
+        _ctx: &RequestContext, // TODO will we ever need it here?
     ) -> anyhow::Result<UninitializedTimeline> {
```
Unused lifetime `'a`?

sharnoff commented on pageserver/src/consumption_metrics.rs:
```diff
@@ -5,13 +5,14 @@
 use anyhow;
+use pageserver_api::models::TenantState;
 use tracing::*;
 use utils::id::NodeId;
 use utils::id::TimelineId;
 
+use crate::context::RequestContext;
 use crate::task_mgr;
 use crate::tenant::mgr;
-use pageserver_api::models::TenantState;
```
Was going to comment about moving `pageserver_api` imports, but... seems like we're pretty inconsistent anyways about whether we group `pageserver_api` with imports from the same crate or external crates.

problame replied:
Yeah, cleaned it up.

sharnoff reviewed (approved):
General idea seems pretty sound - nothing really jumped out as notable other than what I've explicitly commented on.

Normally I'd be concerned about passing a bunch of references around when describing _tasks_, because that can get annoying with spawn requiring `'static` and such, but it seems like you already have the tools needed to make that ok (i.e. `detached_child`).

LizardWizzard commented on pageserver/src/context.rs:
```diff
@@ -0,0 +1,209 @@
+
+    /// Use this function when you have a parent context, the async task you
+    /// are creating matches the description of [`detached_child`], but with
+    /// some additional code changes could be made [`attached_child`].
```
Do not understand the comment. Could you please elaborate? Feel free to resolve if the comment is just a placeholder

problame replied:
Tried to explain what I meant here, but actually, it's just one place where we use this and there, using `detached_child` is OK for now.

Will delete this method.

LizardWizzard commented on pageserver/src/http/routes.rs:
```diff
@@ -179,20 +196,23 @@ async fn timeline_create_handler(mut request: Request<Body>) -> Result<Response<
         .new_timeline_id
         .unwrap_or_else(TimelineId::generate);
 
+    let ctx = RequestContext::new(TaskKind::MgmtRequest, DownloadBehavior::Download);
```
Not sure I fully understand why `timeline_create` can download layers. Should it be `Error` instead?

hlinnaka replied:
Hmm, agreed, timeline creation shouldn't require downloading anything. Let's make it `Error`, or `Warn` if we're note quite sure.

LizardWizzard commented on pageserver/src/http/routes.rs:
```diff
@@ -270,6 +292,8 @@ async fn timeline_detail_handler(request: Request<Body>) -> Result<Response<Body
         query_param_present(&request, "include-non-incremental-logical-size");
     check_permission(&request, Some(tenant_id))?;
 
+    let ctx = RequestContext::new(TaskKind::MgmtRequest, DownloadBehavior::Download);
```
Same here. Though it may require downloads for size maybe? Still would be good to keep `detail` handler free of side effects

hlinnaka replied:
Right, logical size calculation needs downloading

LizardWizzard commented on pageserver/src/pgdatadir_mapping.rs:
```diff
@@ -519,14 +553,15 @@ impl Timeline {
+                .list_rels(spcnode, dbnode, lsn, ctx)
                 .await?
-                .into_iter()
+                .iter()
```
NIT: Is there any reason to moving from `into_iter` to `iter` + `cloned`?

problame replied:
No, just fallout from the extraction.

LizardWizzard commented on pageserver/src/tenant.rs:
```diff
@@ -447,6 +451,7 @@ impl Tenant {
         local_metadata: Option<TimelineMetadata>,
         ancestor: Option<Arc<Timeline>>,
         first_save: bool,
+        _ctx: &RequestContext, /* TODO: do we ever need a context for timeline setup?  */
```
In theory whole tenant can be cancelled during startup, so makes sense to me

LizardWizzard reviewed (commented):
Went through the most of the changes

problame commented on pageserver/src/context.rs:
```diff
@@ -0,0 +1,209 @@
+//!
+//! The solution is that all code paths are infected with precisely one
+//! [`RequestContext`] argument. Functions in the middle of the call chain
+//! only need to pass it on.
```
Thanks :pray:

arssher reviewed (approved):
Haven't thought deeply how context hierarchy exactly can be implemented and what is the cancellation semantics is going to be (e.g. how customizable it is, can the canceled task gracefully shutdown, how waiting for children gonna work, if any), but seems like this solves current problem.

problame commented:
Addressed review comments in latest push. Now rebasing, then force-pushing. The remaining unresolved conversations still require attention, even after the rebase.

problame commented:
Rebase pushed.
Main conflict in `consumption_metrics.rs`.
`origin/main` added a `task_mgr::spawn` call there.
That tasks calculates the synthetic size.
For that, it calculates logical size, and that needs a `ctx` as of this PR.
Solved by creating a `ctx.detached_child()` for the task_mgr task.

What do you think about `DownloadBehavior::Download` for the synthetic size calculation?

```
    // spin up background worker that caclulates tenant sizes
    let worker_ctx =
        ctx.detached_child(TaskKind::CalculateSyntheticSize, DownloadBehavior::Download);
    task_mgr::spawn(
        BACKGROUND_RUNTIME.handle(),
        TaskKind::CalculateSyntheticSize,
```

hlinnaka commented on pageserver/src/bin/pageserver.rs:
```diff
@@ -325,6 +326,10 @@ fn start_pageserver(conf: &'static PageServerConf) -> anyhow::Result<()> {
         if let Some(metric_collection_endpoint) = &conf.metric_collection_endpoint {
+            let metrics_ctx = RequestContext::todo_child(
+                TaskKind::MetricsCollection,
+                DownloadBehavior::Error, // metrics collector shouldn't be downloading anything
```
Calculating the synthetic size requires information about relation sizes, which can require downloading stuff. But I believe this is correct, because we use this context only for the long-running collector, and create child contexts with DownloadBehavior::Download to do the synthetic size calculations. Let's explain that in the comment here, this is misleading as it is.

hlinnaka commented on pageserver/src/context.rs:
```diff
@@ -0,0 +1,198 @@
+//! This module defines `RequestContext`, a structure that we use throughout
+//! the pageserver codebase to propagate high-level context from the places
+//! that _originate_ activity down to the shared code paths at the
+//! heart of the pageserver.
+//!
+//! For example, the initial product needs were that we needed to answer the
+//! following questions inside `Timline::get(page_nr, lsn)`:
+//! 1. What high-level activity ([`TaskKind`]) needs this page?
+//!    We need that information as a categorical dimension for page access
+//!    statistics, which we, in turn, need to guide layer eviction policy design.
+//! 2. How should we behave if, to produce the page image, we need to
+//!    on-demand download a layer file ([`DownloadBehavior`]).
+//!
```
minor copy-editing:
```suggestion
//! This module defines `RequestContext`, a structure that we use throughout
//! the pageserver to propagate high-level context from places
//! that _originate_ activity down to the shared code paths at the
//! heart of the pageserver.
//!
//! For example, in `Timeline::get(page_nr, lsn)` we need to answer the following questions:
//! 1. What high-level activity ([`TaskKind`]) needs this page?
//!    We need that information as a categorical dimension for page access
//!    statistics, which we, in turn, need to guide layer eviction policy design.
//! 2. How should we behave if, to produce the page image, we need to
//!    on-demand download a layer file ([`DownloadBehavior`]).
//!
```

hlinnaka commented on pageserver/src/context.rs:
```diff
@@ -0,0 +1,198 @@
+//! [`RequestContext`] satisfies those needs.
+//! The current implementation is a small `struct` that is passed through
+//! the call chain by reference.
+//!
```
All the text that follows concerns things that are not implemented here yet. When I read this from beginning to end, I didn't realize that until I reached the "We do not yet have a systematic cancellation story" paragraph. Let's add here:

```suggestion
//!
//! # TODO
//!
```

hlinnaka commented on pageserver/src/context.rs:
```diff
@@ -0,0 +1,198 @@
+//! - Request IDs for distributed tracing
+//! - Request/Timeline/Tenant-scoped log levels
+//!
+//! RequestContext might look quite different once it supports those features.
+//! Likely, it will have a shape similar to Golang's `context.Context`.
```
I think this is already pretty similar to Golang's `context.Context`. Maybe mention that at the very beginning already, in the first paragraph in this file.

hlinnaka commented on pageserver/src/context.rs:
```diff
@@ -0,0 +1,198 @@
+//! along the call chain from high level to low level, but few of the functions
+//! in the middle need to understand it.
+//! Further, it is to be expected that we will need to propagate more data
+//! in the future (see the section on future work below).
```
The section on future work is above, not below.

hlinnaka commented on pageserver/src/context.rs:
```diff
@@ -0,0 +1,198 @@
+//! RequestContext might look quite different once it supports those features.
+//! Likely, it will have a shape similar to Golang's `context.Context`.
+//!
+//! ### Why A Container Struct Instead Of Parameters
```
```suggestion
//! ### Why A Context Struct Instead Of Parameters
```
or if you meant container struct, maybe lower case it:

```suggestion
//! ### Why a container struct instead of parameters
```

problame replied:
I meant `Container`, but it's misleading. I'll just omit it.

hlinnaka reviewed (commented):
Reviewed the file comment at top of `context.rs` now. Will continue with the rest later.

problame commented:
There are still unresolved conversations that I don't have a great answer to. @hlinnaka , can you take a look?
Also, are you planning to invest more review time or should we merge this?

Also, I've been thinking about reducing risk of misplaced `DownloadBehavior::Error`. How about, Instead of having `Download`, `Warn`, and `Error`, we just have `Download` and `Error`, and add a config knob / env var to turn `Error` into  what `Warn` currently does, in production, or, `cfg(not(testing))?

problame commented:
Force push: rebased. Minor conflicts in size.rs

hlinnaka reviewed (approved):
> There are still unresolved conversations that I don't have a great answer to. @hlinnaka , can you take a look?
> Also, are you planning to invest more review time or should we merge this?

Replied on the two two unresolved conversations that @LizardWizzard opened. I think this is ready to be merged.

> Also, I've been thinking about reducing risk of misplaced DownloadBehavior::Error. How about, Instead of having Download, Warn, and Error, we just have Download and Error, and add a config knob / env var to turn Error into what Warn currently does, in production, or, `cfg(not(testing))?

Works for me.

If you do that, I would prefer a runtime configuration option rather than build time.

hlinnaka commented:
> > Also, I've been thinking about reducing risk of misplaced DownloadBehavior::Error. How about, Instead of having Download, Warn, and Error, we just have Download and Error, and add a config knob / env var to turn Error into what Warn currently does, in production, or, `cfg(not(testing))?
> 
> Works for me.
> 
> If you do that, I would prefer a runtime configuration option rather than build time.

Hmm, I wonder if we might want to have a operations in the future , where it's expected that you might access a layer that needs downloading, but you would rather get an error instead of performing the download? For example, imagine an opportunistic compaction operation that reads files if they happen to already be present, and compact them, but you don't want to download anything just to compact them. Or we might want to do that in the logical size calculation for the `/timeline_detail` API call, so that we don't force downloading layers just to calculate the logical size, but return it if it can be calculated.

We could easily add a new DownloadBehavior for that later too, though. So up to you.

problame commented:
Asked SRE team about best practice for killswitches: https://neondb.slack.com/archives/C039YKBRZB4/p1674556815184939

koivunej commented on pageserver/src/page_service.rs:
```diff
@@ -589,15 +628,17 @@ impl PageServerHandler {
     }
 
-    #[instrument(skip(self, timeline, req), fields(rel = %req.rel, blkno = %req.blkno, req_lsn = %req.lsn))]
+    #[instrument(skip(self, timeline, req, ctx), fields(rel = %req.rel, blkno = %req.blkno, req_lsn = %req.lsn))]
```
Just FYI there's also a `skip_all` if you are going to skip all, and define custom fields anyway.

problame replied:
Yeah, I tent to do that for new functions, but here, just cargo-culting on and on :)

koivunej reviewed (approved):
No new concerns, looking forward to the next step :) Let's hope the `skip_all` comment was new.

koivunej commented on pageserver/src/http/routes.rs:
```diff
@@ -457,13 +489,14 @@ async fn tenant_size_handler(request: Request<Body>) -> Result<Response<Body>, A
 
     // this can be long operation, it currently is not backed by any request coalescing or similar
     let inputs = tenant
-        .gather_size_inputs()
+        .gather_size_inputs(&ctx)
         .await
         .map_err(ApiError::InternalServerError)?;
```
created a conflict here by merging #3419, not a difficult one unless it get's mangled up.

problame replied:
Thanks for the heads-up!

problame commented:
Force-push: rebased (neglible conflicts) + squashed + improved commit message.

I consider this ready for merge now.

problame commented:
For completeness: I added a killswitch that's configurable via pageserver.toml. Will work with SRE to create a workflow that can flip the kill switch in <10min.

koivunej commented:
Post-conflict resolution `cargo fmt --all` required.

problame commented:
Force-push: rebase due to conflict in task_mgr.rs comments (my own PR that merged earlier today)

problame commented:
Force-push: rebase.
Minor conflict with
https://github.com/neondatabase/neon/commit/572332ab50b2d87fa15b11b2c8c1f9e1313c915b

Enabling auto-merge now

koivunej commented on pageserver/src/tenant/timeline.rs:
```diff
@@ -1312,8 +1369,18 @@ impl Timeline {
             .get()
             .is_none());
+
+        info!(
+            "spawning logical size computation from context of task kind {:?}",
+            ctx.task_kind()
+        );
```
Only now ran into this, was this intentional? I could remove it on #3431..

problame replied:
It was kind of intentional, because logical size computation is triggered from multiple places, and will likely be a cause of unwanted layer downloads.

koivunej replied:
Just occured to me that this could be solved 100% with spans. I'll demo this with #3431.
